src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp

```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
+        const auto * loaded_descriptor = registry_pool()->FindFileByName(std::to_string(id));
+        if (loaded_descriptor)
+            return loaded_descriptor;
+
+        return fetchSchema(id);
+    }
+
+    const google::protobuf::FileDescriptor* fetchSchema(uint32_t id)
+    {
+        auto schema = registry.fetchSchema(id);
+        std::string schema_content {schema.data(), schema.size()};
+        google::protobuf::io::ArrayInputStream input{schema.data(), static_cast<int>(schema.size())};
+        google::protobuf::io::Tokenizer tokenizer(&input, this);
+        google::protobuf::FileDescriptorProto descriptor;
+        descriptor.set_name(std::to_string(id));
+        google::protobuf::compiler::Parser parser;
+        parser.RecordErrorsTo(this);
+        parser.Parse(&tokenizer, &descriptor);
+
+        auto const * ret = registry_pool()->BuildFile(descriptor);
+        auto mt_count = ret->message_type_count();
+        if (mt_count < 1)
+            throw Exception(ErrorCodes::INVALID_DATA, "No message type in schema");
+        return ret;
+    }
+
+    KafkaSchemaRegistry registry;
+    google::protobuf::DescriptorPool registry_pool_;
+
+    google::protobuf::DescriptorPool* registry_pool() { return &registry_pool_; }
+};
+
+ProtobufConfluentRowInputFormat::ProtobufConfluentRowInputFormat(
+    ReadBuffer & in_, const Block & header_, Params params_, const FormatSettings &  format_settings_)
+    : IRowInputFormat(header_, in_, params_, ProcessorID::ProtobufRowInputFormatID)
+    , registry(getConfluentSchemaRegistry(format_settings_.schema.kafka_schema_registry_url, format_settings_.schema.kafka_schema_registry_credentials))
+{
+}
+
+bool ProtobufConfluentRowInputFormat::readRow(MutableColumns & columns, RowReadExtension & row_read_extension)
+{
+    if (in->eof())
+        return false;
+
+    auto schema_id = KafkaSchemaRegistry::readSchemaId(*in);
+
+
+    Int64 indexes_count = 0;
+    readVarInt(indexes_count, *in);
+
+    std::vector<Int64> indexes;
+    indexes.reserve(indexes_count < 1 ? 1 : indexes_count);
+
+    if (indexes_count < 1)
+        indexes.push_back(0);
+    else
+    {
+        for (size_t i = 0; i < static_cast<size_t>(indexes_count); i++)
+        {
+            Int64 ind = 0;
+            readVarInt(ind, *in);
+            indexes.push_back(ind);
+        }
+    }
+
+    const auto & header = getPort().getHeader();
+
+    ProtobufReader reader {*in};
+    serializer = ProtobufSerializer::create(
+        header.getNames(),
+        header.getDataTypes(),
+        missing_column_indices,
+        *registry->getMessageType(schema_id, indexes),
+        /*with_length_delimiter=*/false,
+        reader);
+
+    size_t row_num = columns.empty() ? 0 : columns[0]->size();
+    if (!row_num)
+        serializer->setColumns(columns.data(), columns.size());
+
+    serializer->readRow(row_num);
+
+    row_read_extension.read_columns.clear();
+    row_read_extension.read_columns.resize(columns.size(), true);
```

Comment:
if missing_column_indices is zero, i was thinking we can optimize this as  following ? and i recalled the framework (parent driving template function) does the clear before the call

```
assert(row_read_extension.read_columns.empty());
if (!missing_column_indices.empty())
{
    row_read_extension.read_columns.resize(columns.size(), true);
     for (size_t column_idx : missing_column_indices)
        row_read_extension.read_columns[column_idx] = false;
}

```